grants/speculative/Dkg_Multisig_Wallet.md

```diff
@@ -0,0 +1,152 @@
+
+## Team Code Repos
+* https://github.com/everstake
+* https://github.com/atticlab
+
+## Team LinkedIn Profiles
+* https://www.linkedin.com/in/vitalii-parkhomenko-493561187/
+* https://www.linkedin.com/in/aleksei-korobeinikov/
+* https://www.linkedin.com/in/pavelmidvel/
+* https://www.linkedin.com/in/stanislav-c-362685195/
+
+## Development Roadmap
+1. Wallet APP
+   * Wallet application (160h)
+      * Implement UI and layouts - 44h
+      * Write main logic and working with adapter - 100h
+         * ShowWalletStatus
+         * CreateWallet
+         * JoinWallet
+         * GenerateKey
+         * CreateTransaction
+         * DoBackUp
+       * Test UX of application - 4h
+       * Do manual testing - 4h
+       * Write unit tests - 8h
+   * Adapter in Rust (200h)
+      * Write all the methods for external calls - 168h
+         * NewDistKeyGen
+         * ProcessDeal
+         * ProcessResponse
+         * ProcessSecretCommits
+         * DistKeyShare
+         * NewDSS
+         * ProcessPartialSignature
+         * ConstructFinalSignature
+         * VerifySignature
+      * Integrate adapter with DKG library - 8h
+      * Write unit tests for Rust code - 16h
+      * Write unit tests for new compiled JS code - 8h
+   * Integration testing adapter with application (16h)
+   * Integration testing application with server (8h)
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Total for Wallet App: 384h - 1 back-end and 1 front-end developers (2 months)
+
+2. Web server
+   * API written with actix
+      * Write all the endpoints with asynchronous functionality - 124h
+         * NewWallet
+         * JoinWallet
+         * BroadcastDeal
+         * CreateNewTransaction
+         * BroadcastPartialSignature
+         * BroadcastFinalSignature
+      * Write functionality for authentication and encryption - 16h
+         * NewUser
+         * CheckUser
+         * Encrypt
+         * Decrypt
+      * Integrate functionality with DB - 4h
+      * Write unit tests - 16h
+      * Configure all this with Docker to easier deploy it to server - 8h
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Total for Web server: 168h - 1 back-end developer (1 month)
+
+3. DKG library
+   * VSS (240h)
+      * Implement polynomial commitments with Lagrange interpolation - 72h
+         * NewPrivPoly
+         * RecoverSecret
+         * NewPubPoly
+      * Implement Dealer functionality (struct which will create shares and deals(set of data such polynomials, sessionID, threshold)) - 56h
+         * NewDealer
+         * EncryptDeal
+         * SecretCommit
+      * Implement Verifier functionality (struct which will receive deals from all the participants and verify it) - 96h
+         * NewVerifier
+         * ProcessEncryptedDeal
+      * Write unit tests - 16h
```

Comment:
I've mentioned this before, but [`schnorrkel::Keypair::dleq_prove`](https://docs.rs/schnorrkel/0.9.1/schnorrkel/keys/struct.Keypair.html#method.dleq_proove) should help with VSS.  We could just add VSS directly to schnorrkel, but that's not essential probably.

Reply:
Probably there’s a misunderstanding here.
By the “Verifier functionality” we mean the struct which will create encrypted Deals, first it generates a temporary Diffie Hellman key, signs it using its long-term key, and computes the shared key depending on its long-term and ephemeral key and the verifier's public key. This shared key is then fed into a HKDF whose output is the key to an AEAD (AES256-GCM) scheme to encrypt the deal. Then to verify this Deal receiver should decrypt it and verify its shares against the public coefficients.